File: benchmark/languages/go/work/day05/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Range struct {
	lo, hi int
}

func parseRange(line string) (Range, bool) {
	var lo, hi int
	// Try format "lo-hi"
	n, err := fmt.Sscanf(line, "%d-%d", &lo, &hi)
	if err == nil && n == 2 {
		if lo > hi {
			lo, hi = hi, lo
		}
		return Range{lo, hi}, true
	}
	return Range{}, false
}

func parseInt(line string) (int, bool) {
	n, err := strconv.Atoi(strings.TrimSpace(line))
	if err != nil {
		return 0, false
	}
	return n, true
}

func parseInput(filename string) ([]Range, []Range, error) {
	f, err := os.Open(filename)
	if err != nil {
		return nil, nil, err
	}
	defer f.Close()

	var freshRanges []Range
	// Second section: either individual IDs (stored as single-element ranges) or ranges
	var availRanges []Range
	inSecond := false

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1024*1024), 1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			inSecond = true
			continue
		}

		if !inSecond {
			r, ok := parseRange(line)
			if !ok {
				return nil, nil, fmt.Errorf("bad fresh range: %q", line)
			}
			freshRanges = append(freshRanges, r)
		} else {
			// Try as range first
			if r, ok := parseRange(line); ok {
				availRanges = append(availRanges, r)
			} else if n, ok := parseInt(line); ok {
				// Single integer
				availRanges = append(availRanges, Range{n, n})
			} else {
				return nil, nil, fmt.Errorf("bad available item: %q", line)
			}
		}
	}

	return freshRanges, availRanges, scanner.Err()
}

// mergeRanges merges overlapping ranges and returns the merged list sorted by lo.
func mergeRanges(ranges []Range) []Range {
	if len(ranges) == 0 {
		return nil
	}
	sorted := make([]Range, len(ranges))
	copy(sorted, ranges)
	sort.Slice(sorted, func(i, j int) bool {
		if sorted[i].lo != sorted[j].lo {
			return sorted[i].lo < sorted[j].lo
		}
		return sorted[i].hi < sorted[j].hi
	})

	merged := []Range{sorted[0]}
	for _, r := range sorted[1:] {
		last := &merged[len(merged)-1]
		if r.lo <= last.hi+1 {
			if r.hi > last.hi {
				last.hi = r.hi
			}
		} else {
			merged = append(merged, r)
		}
	}
	return merged
}

// isFreshBinary uses binary search on merged (sorted) ranges
func isFreshBinary(id int, merged []Range) bool {
	lo, hi := 0, len(merged)-1
	for lo <= hi {
		mid := (lo + hi) / 2
		r := merged[mid]
		if id < r.lo {
			hi = mid - 1
		} else if id > r.hi {
			lo = mid + 1
		} else {
			return true
		}
	}
	return false
}

// countFreshInRange counts how many integers in [qlo, qhi] are in merged fresh ranges
func countFreshInRange(qlo, qhi int, merged []Range) int {
	// Find all merged ranges that overlap with [qlo, qhi] and sum
	count := 0
	for _, r := range merged {
		if r.hi < qlo || r.lo > qhi {
			continue
		}
		lo := r.lo
		if lo < qlo {
			lo = qlo
		}
		hi := r.hi
		if hi > qhi {
			hi = qhi
		}
		count += hi - lo + 1
	}
	return count
}

func part1(freshRanges []Range, availRanges []Range) int {
	merged := mergeRanges(freshRanges)

	// Check each available item
	// If single ID (lo==hi), check if fresh
	// If range, count how many in that range are fresh
	count := 0
	for _, ar := range availRanges {
		if ar.lo == ar.hi {
			// Single ID
			if isFreshBinary(ar.lo, merged) {
				count++
			}
		} else {
			// Range of IDs — count fresh ones
			count += countFreshInRange(ar.lo, ar.hi, merged)
		}
	}
	return count
}

func part2(freshRanges []Range) int {
	merged := mergeRanges(freshRanges)
	total := 0
	for _, r := range merged {
		total += r.hi - r.lo + 1
	}
	return total
}

func main() {
	freshRanges, availRanges, err := parseInput("input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}

	part := ""
	if len(os.Args) > 1 {
		part = os.Args[1]
	}

	switch part {
	case "1":
		fmt.Printf("Part 1: %d\n", part1(freshRanges, availRanges))
	case "2":
		fmt.Printf("Part 2: %d\n", part2(freshRanges))
	default:
		fmt.Printf("Part 1: %d\n", part1(freshRanges, availRanges))
		fmt.Printf("Part 2: %d\n", part2(freshRanges))
	}
}

File: benchmark/languages/c/work/day05/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    long long lo;
    long long hi;
} Range;

int cmp_range(const void *a, const void *b) {
    const Range *ra = (const Range *)a;
    const Range *rb = (const Range *)b;
    if (ra->lo < rb->lo) return -1;
    if (ra->lo > rb->lo) return  1;
    if (ra->hi < rb->hi) return -1;
    if (ra->hi > rb->hi) return  1;
    return 0;
}

int main(int argc, char *argv[]) {
    int do_part1 = 1, do_part2 = 1;
    if (argc >= 2) {
        if (argv[1][0] == '1') do_part2 = 0;
        else if (argv[1][0] == '2') do_part1 = 0;
    }

    FILE *f = fopen("input.txt", "r");
    if (!f) { fprintf(stderr, "Cannot open input.txt\n"); return 1; }

    // Read all lines
    char line[256];
    int in_ranges = 1;  // start in range section

    int ranges_cap = 64;
    int ranges_count = 0;
    Range *ranges = malloc(ranges_cap * sizeof(Range));

    int ids_cap = 64;
    int ids_count = 0;
    long long *ids = malloc(ids_cap * sizeof(long long));

    while (fgets(line, sizeof(line), f)) {
        // Trim newline
        int len = (int)strlen(line);
        while (len > 0 && (line[len-1] == '\n' || line[len-1] == '\r'))
            line[--len] = '\0';

        if (len == 0) {
            // blank line -> switch to IDs section
            in_ranges = 0;
            continue;
        }

        if (in_ranges) {
            // parse "lo-hi"
            long long lo, hi;
            if (sscanf(line, "%lld-%lld", &lo, &hi) == 2) {
                if (ranges_count == ranges_cap) {
                    ranges_cap *= 2;
                    ranges = realloc(ranges, ranges_cap * sizeof(Range));
                }
                ranges[ranges_count].lo = lo;
                ranges[ranges_count].hi = hi;
                ranges_count++;
            }
        } else {
            // parse ID
            long long id;
            if (sscanf(line, "%lld", &id) == 1) {
                if (ids_count == ids_cap) {
                    ids_cap *= 2;
                    ids = realloc(ids, ids_cap * sizeof(long long));
                }
                ids[ids_count++] = id;
            }
        }
    }
    fclose(f);

    // Sort ranges for efficient lookup and merging
    qsort(ranges, ranges_count, sizeof(Range), cmp_range);

    // --- Part 1: count how many IDs are in any range ---
    long long part1 = 0;
    for (int i = 0; i < ids_count; i++) {
        long long id = ids[i];
        // Binary search for a range that could contain id
        int lo_idx = 0, hi_idx = ranges_count - 1, found = 0;
        // Find the last range with lo <= id
        int best = -1;
        while (lo_idx <= hi_idx) {
            int mid = (lo_idx + hi_idx) / 2;
            if (ranges[mid].lo <= id) {
                best = mid;
                lo_idx = mid + 1;
            } else {
                hi_idx = mid - 1;
            }
        }
        if (best >= 0 && ranges[best].hi >= id) {
            found = 1;
        }
        // Also check if there's an overlapping range that starts after id
        // but that can't contain id if lo > id, so no need.
        // However there might be multiple ranges with lo <= id, best is the last one.
        // But wait: we only checked the last range with lo<=id. We should check all ranges
        // with lo<=id to see if any has hi>=id. Since ranges are sorted by lo,
        // we can just check up to 'best' — but any range before 'best' with hi>=id also works.
        // Actually the best approach: find all ranges with lo<=id, check max hi among them.
        // We can precompute prefix max hi, but simpler: just scan backward from best.
        if (!found && best >= 0) {
            // check earlier ranges (they have smaller lo, might have large hi)
            for (int j = best - 1; j >= 0; j--) {
                if (ranges[j].hi >= id) { found = 1; break; }
                // since ranges[j].lo <= ranges[best].lo <= id,
                // if ranges[j].hi < id, earlier ones might still have hi >= id
                // we must keep checking
            }
        }
        if (found) part1++;
    }

    // --- Part 2: merge ranges and count total IDs ---
    // Build merged ranges
    int merged_cap = ranges_count + 1;
    Range *merged = malloc(merged_cap * sizeof(Range));
    int merged_count = 0;

    for (int i = 0; i < ranges_count; i++) {
        if (merged_count == 0) {
            merged[merged_count++] = ranges[i];
        } else {
            Range *last = &merged[merged_count - 1];
            if (ranges[i].lo <= last->hi + 1) {
                // overlapping or adjacent: extend
                if (ranges[i].hi > last->hi) last->hi = ranges[i].hi;
            } else {
                merged[merged_count++] = ranges[i];
            }
        }
    }

    long long part2 = 0;
    for (int i = 0; i < merged_count; i++) {
        part2 += merged[i].hi - merged[i].lo + 1;
    }

    if (do_part1) printf("Part 1: %lld\n", part1);
    if (do_part2) printf("Part 2: %lld\n", part2);

    free(ranges);
    free(ids);
    free(merged);
    return 0;
}
